frontend/src/app/components/MainInstallButton/index.js

```diff
@@ -26,62 +26,34 @@ export default class MainInstallButton extends React.Component {
     };
   }
 
-  enableExperimentWrapped() {
-    const { experiment, enableExperiment, postInstallCallback } = this.props;
-
-    this.setState({ isInstalling: true });
-    return enableExperiment(experiment)
-      .then(() => {
-        if (postInstallCallback) {
-          postInstallCallback();
-        }
-      }).catch((err) => {
-        this.setState({isInstalling: false});
-      });
-  }
-
   install(e: Object) {
     // Don't install if a mouse button other than the primary button was clicked
     if (e.button !== 0) {
       return;
     }
-    const { requireRestart, sendToGA, eventCategory,
-      installAddon, installCallback, navigateTo, eventLabel,
-      postInstallCallback, isExperimentEnabled, hasAddon,
-      experiment, experimentTitle } = this.props;
-
-    if (hasAddon) {
-      if (!isExperimentEnabled(experiment)) {
-        this.enableExperimentWrapped();
-      } else navigateTo("/experiments");
-      return;
-    }
+
+    const { sendToGA, eventCategory, enableExperiment,
+      installAddon, installCallback, eventLabel,
+      experiment } = this.props;
+
+    this.setState({ isInstalling: true });
 
     if (installCallback) {
-      this.setState({ isInstalling: true });
       installCallback(e);
-      return;
-    }
+    } else {
+      const install = experiment ?
+        enableExperiment(experiment, eventCategory, eventLabel) :
+        installAddon(sendToGA, eventCategory, eventLabel);
 
-    this.setState({ isInstalling: true },
-      () => installAddon(requireRestart, sendToGA, eventCategory, eventLabel)
-        .then(() => {
-          // one click install and enable handling
-          if (experimentTitle) {
-            return this.enableExperimentWrapped();
-          } else if (postInstallCallback) {
-            postInstallCallback();
-          }
-          return Promise.resolve();
-        })
-        .catch(err => {
-          this.setState({ isInstalling: false });
-        }));
+      install.catch(err => {
+        this.setState({ isInstalling: false });
+      });
+    }
   }
 
   render() {
-    const { isFirefox, isMinFirefox, isMobile, hasAddon, experimentTitle, experimentLegalLink } = this.props;
-    const { isInstalling } = this.state;
+    const { isFirefox, isMinFirefox, isMobile, hasAddon, experimentTitle, experimentLegalLink, experiment } = this.props;
+    const isInstalling = this.state.isInstalling || (experiment && experiment.inProgress);
```

Comment:
This ensures the button remains in the installing state during the whole one-click flow. The react cycle is slightly different from how it worked before because we transition from HomePageNoAddon to HomePageWithAddon in the middle now.